### PR TITLE
WIP: add analysis run engine

### PR DESCRIPTION
**Pull request by CJ-Wright**
This is still a work in progress

### CODE REVIEW

**Comment by CJ-Wright** [comment]
Closing to reopen and restart travis

**Comment by CJ-Wright** [comment]
@chiahaoliu @sbanerjee23 @sbillinge This is ready for review.

File: redsky/analysis_run_engine.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+        """
+        if not isinstance(hdrs, list):
+            hdrs = [hdrs]
+        # issue run start
+        run_start_uid = self.an_db.mds.insert_run_start(
+            uid=str(uuid4()), time=time.time(),
+            parents=[hdr['start']['uid'] for hdr in hdrs],
+            function_name=run_function.__name__,
+            provenance={'args': args,
```
Comment (CJ-Wright) [inline redsky/analysis_run_engine.py]: Most likely this chunk will store the error and not the environment. I think having a hard dependency on conda is not too unreasonable.

**Comment by CJ-Wright** [comment]
@tacaswell @danielballan if you get a second can you take a look at this?

File: redsky/analysis_run_engine.py (hunk 1 above)
Comment (tacaswell) [inline redsky/analysis_run_engine.py]: Having a hard dep in conda is not all that reasonable, but you can make it pluggable later.

Comment (CJ-Wright) [inline redsky/analysis_run_engine.py]: Is there a more complete way to capture the environment? I'm borrowing this chunk from one of the prior analysis store schemas.

**Comment by tacaswell** [comment]
Hard-coding the inserts into mds/fs I think is a bit of an anti-pattern as is assuming that the source databroker and the destination databroker are the same.   Probably better to start with taking in the event streams.

I suggest embracing dicts over 'happen to be aligned' lists.

Instead of doing a frame-work thing first, I strongly suggest just manually writing out the analysis store integration, it will probably be less work than manually marking up every function you want to use with a sufficient amount of metadata.  Once this has been done at a couple of beamlines, we might know enough to start generalizing.

Also suggest embracing python3 with annotation + signature.

Have a look at https://github.com/qiime2/qiime2 which is much farther along at automatically marking up analysis functions (but from their scipy talk seem very tied to _their_ version of data artifacts).

The core of what you have here (wrapping functions in objects, using a higher-layer executor to run the wrapped functions, optionally serialize the results) is the core of what vistrails does.

There does not seem to be a way to chain multiple steps together?

See https://github.com/BrookhavenNationalLaboratory/pyRafters for my abandoned attempt at something like this.

[ran out of time on the train]

**Comment by CJ-Wright** [comment]
@tacaswell Thank you very much for your through review (I shudder to think what would have happened if you didn't run out of time).

I think I will try to address your comments piecemeal since each one deserves quite a bit of thought.

Under the current code one could simply hand the output header from one analysis run engine to another to chain the two together. In the long run I personally would like to use dask, see #2. I have very little interest in performing the scheduling/task graph assembling/optimization myself. Ideally I would like this system to just act as a passive observer, tracking what went in and what went out with enough metadata to recreate things (if possible). The similar logic applies to vistrails, I would hope that this could just observe what is going on without getting into the problem of how users assemble their workflows.

**Comment by CJ-Wright** [comment]
I guess I am a bit uncertain about "I suggest embracing dicts over 'happen to be aligned' lists." what do you mean?

**Comment by CJ-Wright** [comment]
I will definitely start using annotations for the functions, is it ok to use them as info for the documents?

**Pull request closed without merging**